group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java

```diff
@@ -5524,5 +5670,2897 @@ private static Record newGroupMetadataRecord(
+            .withGroupInstanceId("follower-instance-id")
+            .withMemberId(UNKNOWN_MEMBER_ID)
+            .withProtocols(protocols)
+            .withRebalanceTimeoutMs(rebalanceTimeoutMs)
+            .withSessionTimeoutMs(sessionTimeoutMs)
+            .build();
+
+        CompletableFuture<JoinGroupResponseData> responseFuture = new CompletableFuture<>();
+        CoordinatorResult<Void, Record> result = context.sendGenericGroupJoin(request, responseFuture);
+        assertTrue(result.records().isEmpty());
+        assertFalse(responseFuture.isDone());
+        assertEquals("roundrobin", group.selectProtocol());
+
+        // Old leader hasn't joined in the meantime, triggering a re-election.
+        context.sleepAndAssertEmptyResult(rebalanceTimeoutMs);
+        assertTrue(responseFuture.isDone());
+        assertEquals(Errors.NONE.code(), responseFuture.get().errorCode());
+        assertTrue(group.hasStaticMember("leader-instance-id"));
+        assertTrue(group.isLeader(responseFuture.get().memberId()));
+        assertNotEquals(rebalanceResult.followerId, responseFuture.get().memberId());
+
+        Set<String> expectedGroupInstanceIds = new HashSet<>();
+        expectedGroupInstanceIds.add("leader-instance-id");
+        expectedGroupInstanceIds.add("follower-instance-id");
+
+        JoinGroupResponseData expectedResponse = new JoinGroupResponseData()
+            .setErrorCode(Errors.NONE.code())
+            .setGenerationId(rebalanceResult.generationId + 1)
+            .setMemberId(responseFuture.get().memberId())
+            .setLeader(responseFuture.get().memberId())
+            .setProtocolName("roundrobin")
+            .setProtocolType("consumer")
+            .setMembers(toJoinResponseMembers(group));
+
+        checkJoinGroupResponse(
+            expectedResponse,
+            responseFuture.get(),
+            group,
+            COMPLETING_REBALANCE,
+            expectedGroupInstanceIds
+        );
+    }
+
+    @Test
+    public void testStaticMemberRejoinWithUnknownMemberIdAndChangeOfProtocolWhileSelectProtocolUnchangedPersistenceFailure()
+        throws Exception {
+
+        GroupMetadataManagerTestContext context = new GroupMetadataManagerTestContext.Builder()
+            .build();
+        GenericGroup group = context.createGenericGroup("group-id");
+
+        RebalanceResult rebalanceResult = staticMembersJoinAndRebalance(
+            context,
+            "group-id",
+            "leader-instance-id",
+            "follower-instance-id"
+        );
+
+        String selectedProtocol = group.selectProtocol();
+        JoinGroupRequestProtocolCollection protocols = new JoinGroupRequestProtocolCollection(0);
+        protocols.add(new JoinGroupRequestProtocol()
+            .setName(selectedProtocol)
+            .setMetadata(ConsumerProtocol.serializeSubscription(new ConsumerPartitionAssignor.Subscription(
+                Collections.singletonList("foo"))).array())
+        );
```

Comment:
This code is repeated a few times. Should we define an helper method to create the protocol from a name and the subscriptions?